Update caching docs (#6547)
apollographql/apollo-server: docs/source/performance/cache-backends.mdx

---
title: Configuring cache backends
description: How to configure Apollo Server's cache
---

# Configuring cache backends

## Introduction

Several features of Apollo Server (such as [automatic persisted queries](./performance/apq), the [response cache plugin](./performance/caching#caching-with-responsecacheplugin-advanced), and [`RESTDataSource`](./data/data-sources#restdatasource-reference)) take advantage of a cache backend. By default, Apollo Server uses a cache in memory but this can be configured to use a different backend such as Redis or Memcached.

> You can specify a cache backend using the `cache` option to the Apollo Server constructor. The cache backend must implement the [`KeyValueCache`](https://github.com/apollographql/apollo-utils/tree/main/packages/keyValueCache#keyvaluecache-interface) interface from the `@apollo/utils.keyvaluecache` package. Apollo maintains an `InMemoryLRUCache` implementation in that package as well as a [`KeyvAdapter`](https://github.com/apollographql/apollo-utils/tree/main/packages/keyvAdapter#keyvadapter-class) class from the `@apollo/utils.keyvadapter` package for wrapping the [`keyv`](https://www.npmjs.com/package/keyv) package which implements a number of useful cache backends.

Author: Good eye, I actually did add this on a separate PR (https://github.com/apollographql/apollo-server/pull/6536) if you wouldn't mind adding feedback over there or just DM'ing it to me. Right now that work only exists on the `release-3.9.0` branch.

Unfortunately, the default configuration of Apollo Server exposes you to denial of service attacks.

> This is because APQs are enabled by default and the default cache in Apollo Server 3 is unbounded, meaning an attacker can exhaust your memory and crash your server. The default cache in Apollo Server 4 will be bounded; **we recommend opting in to that behavior by providing the `cache: "bounded"` option to your Apollo Server constructor or configuring the `cache` yourself.**

Author: [0872021](https://github.com/apollographql/apollo-server/pull/6547/commits/0872021fd289f471b03ad5e0a40610f75a9fad63)

## Configuring in-memory caching

> By default, Apollo Server's caching features use an unbounded cache. This is not safe for production use. If you'd like to configure the in-memory cache, Apollo provides the [`InMemoryLRUCache`](https://github.com/apollographql/apollo-utils/tree/main/packages/keyValueCache#inmemorylrucache) class from the `@apollo/utils.keyvaluecache` package. This class is a wrapper around the `lru-cache` package with a default maximum of approximately 30MiB of memory. The `InMemoryLRUCache` is configurable with the options from `lru-cache`. See the [`lru-cache` documentation`](https://www.npmjs.com/package/lru-cache) for more information on configuration options.

Reviewer: How would you feel about splitting this up into two paragraphs? Something like this? 
```suggestion
**Apollo Server's default caching features use an unbounded cache, which is not safe for production use.** If you'd like to configure the in-memory cache, Apollo provides the [`InMemoryLRUCache`](https://github.com/apollographql/apollo-utils/tree/main/packages/keyValueCache#inmemorylrucache) class from the `@apollo/utils.keyvaluecache` package. 

The `InMemoryLRUCache` class is a wrapper around the `lru-cache` package and has a default maximum of approximately 30MiB of memory. You can configure an instance of `InMemoryLRUCache` with the same options as found in the `lru-cache` package, see the [`lru-cache` documentation](https://www.npmjs.com/package/lru-cache) for more details. 
```

Reviewer: I feel like you should probably specifically say something along the lines of "configuring your cache will make it safe to use in production", maybe adding a specific callout to `cache: "bounded"` in this section as well?

> ### Example: using `InMemoryLRUCache`

Reviewer: I feel like these "Example" headers could probably be removed?  I think the previous "Configuring in-memory caching" buckets both of these examples nicely already.  

You could then add a sentence above this example saying something like "You can configure an `ApolloServer` instance with an  `InMemoryLRUCache`, like so:"

Author: [ae01541](https://github.com/apollographql/apollo-server/pull/6547/commits/ae01541899df5b3761ef8ee1a0134f557dd88c2a)

```ts
import { InMemoryLRUCache } from '@apollo/utils.keyvaluecache';

const server = new ApolloServer({
  // Note: this is equivalent to providing `cache: "bounded"`
  cache: new InMemoryLRUCache(),
});
```

> ### Example: configuring `InMemoryLRUCache`

Reviewer: See comment above about removing "Example" header

Author: [ae01541](https://github.com/apollographql/apollo-server/pull/6547/commits/ae01541899df5b3761ef8ee1a0134f557dd88c2a)

> In this example, we've increased the default size and provided a default TTL. For more information on the options, see the [`lru-cache` documentation`](https://www.npmjs.com/package/lru-cache).

Reviewer: ```suggestion
In this example, we've increased the default size and provided a default TTL. For more information on these configuration options, see the [`lru-cache` documentation`](https://www.npmjs.com/package/lru-cache).
```

```ts
import { InMemoryLRUCache } from '@apollo/utils.keyvaluecache';

const server = new ApolloServer({
  // ...
  cache: new InMemoryLRUCache({
    // ~100MiB
    maxSize: Math.pow(2, 20) * 100,
    // 5 minutes
    ttl: 300000,
  }),
});
```

## Configuring external caching

Apollo no longer maintains any caching backends directly. Instead, we recommend using the [`keyv`](https://www.npmjs.com/package/keyv) package along with the [`KeyvAdapter`](https://github.com/apollographql/apollo-utils/tree/main/packages/keyvAdapter#keyvadapter-class) class provided by the `@apollo/utils.keyvadapter` package. `KeyvAdapter` simply wraps a `Keyv` instance and implements the `KeyValueCache` interface which is required by Apollo Server.

### Example: using `Keyv` with Redis

```ts
import Keyv from 'keyv';
import { KeyvAdapter } from '@apollo/utils.keyvadapter';

const server = new ApolloServer({
  // ...,
  cache: new KeyvAdapter(new Keyv('redis://localhost:6379')),
});
```

## Implementing your own cache backend

> If your requirements are more specialized or you prefer to implement your own cache backend, you can implement the `KeyValueCache` interface and pass it to Apollo Server's constructor.

Reviewer: ```suggestion
If your requirements are specialized or you'd prefer to implement your own cache backend, you can implement the `KeyValueCache` interface and pass it to the `ApolloServer` constructor directly.
```

> The interface is very simple:

Reviewer: I feel like clarifying what the below snippet is showing is helpful. So, something like: 
```suggestion
The `KeyValueCache` interface is shown below:
```

```ts
interface KeyValueCache<V> {
  get(key: string): Promise<V | undefined>;
  // ttl is specified in seconds
  set(key: string, value: V, options?: { ttl?: number | null }): Promise<void>;
  delete(key: string): Promise<boolean | void>;
}
```

## Legacy caching implementation

> Apollo Server < 3.9.0 used the `apollo-server-caching` package to implement caching. This package is no longer maintained and is not recommended for use. The `KeyValueCache` interface has been moved to the `@apollo/utils.keyvaluecache` package. The `InMemoryLRUCache` class has been moved there as well, however it now uses version 7 of `lru-cache` which accepts different options and no longer allows a cache to be unbounded. The `apollo-server-cache-redis` and `apollo-server-cache-memcached` packages are no longer being published; we recommend using `keyv` instead.

Reviewer: This is a lot of great info, and I think splitting it up will make it more digestible. 
```suggestion
Before Apollo Server 3.9+, we used the `apollo-server-caching` package to implement caching. The `apollo-server-caching` package is no longer maintained, and we do not recommend using it. The `KeyValueCache` interface has been moved and is now in the `@apollo/utils.keyvaluecache` package. 

The `InMemoryLRUCache` class has also moved to the `@apollo/utils.keyvaluecache` package. The `InMemoryLRUCache` class now uses version 7 of `lru-cache`, accepting different configuration options and no longer allowing a cache to be unbounded.

The `apollo-server-cache-redis` and `apollo-server-cache-memcached` packages are no longer published; we recommend using `keyv` instead.
```

---
title: Configuring cache backends
description: How to configure Apollo Server's cache
---

> > ⚠️ New in Apollo Server 3.9: we **highly recommend** that all users pass `cache: "bounded"` or configure their cache in a manner which isn't unbounded (which is current default behavior) to protect your server from attacks which exhaust available memory causing a DOS. Continue reading for more information on configuring Apollo Server's cache.

Reviewer 2: is it reasonable to suggest disabling APQ as another approach here? or can this link to a place that has a bit more details about the issue that could mention that?

Author: I linked to the section immediately below which goes in to more details including APQ cache

> > ⚠️ New in Apollo Server 3.9: we **highly recommend** that all users pass `cache: "bounded"` or configure their cache in a manner which isn't unbounded (which is current default behavior) to protect your server from attacks which exhaust available memory causing a DOS. Continue reading for more information on configuring Apollo Server's cache.

Reviewer 3: ```suggestion
> ⚠️ **New in Apollo Server 3.9:** we _strongly recommend_ that all users pass `cache: "bounded"` or configure their cache in a manner that isn't unbounded (which is current default behavior). This protects your server from attacks that exhaust available memory, causing a DOS. Continue reading for more information on configuring Apollo Server's cache.
```

Author: Wasn't able to directly commit, but added via b30d11eac

> Apollo Server has several features that take advantage of a cache backend (e.g., [automatic persisted queries](./performance/apq), the [response cache plugin](./performance/caching#caching-with-responsecacheplugin-advanced), and [`RESTDataSource`](./data/data-sources#restdatasource-reference)). By default, Apollo Server uses a cache in memory, but you can configure it to use a different backend such as Redis or Memcached.

Reviewer 3: ```suggestion
Many Apollo Server features take advantage of a cache backend (these features include [automatic persisted queries](./performance/apq), the [response cache plugin](./performance/caching#caching-with-responsecacheplugin-advanced), and [`RESTDataSource`](./data/data-sources#restdatasource-reference)). Apollo Server uses an in-memory cache by default, but you can configure it to use a different backend, such as Redis or Memcached.
```

> You can specify a cache backend by passing a `cache` option to the `ApolloServer` constructor. Your specified cache backend _must_ implement the [`KeyValueCache`](https://github.com/apollographql/apollo-utils/tree/main/packages/keyValueCache#keyvaluecache-interface) interface from the `@apollo/utils.keyvaluecache` package.

Reviewer 2: `_must_ implement` seems kinda aggressive; maybe just `implements` or `should implement` or even just not italics? Unless this was a suggestion from Docs

Reviewer: Haha, that was a suggestion from me for added emphasis, I don't feel strongly one way or another though!

Author: Will remove italics. "implements" sounds a little weird, "should implement" is a lot less correct than "must implement".

> There are many cache backend implementations to choose from, including several implementations wrapped by Apollo. For example, Apollo maintains an implementation of `InMemoryLRUCache` in the `@apollo/utils.keyvaluecache` package. Apollo also provides a wrapper class for the  [`keyv` package](https://www.npmjs.com/package/keyv) (which implements several cache backends) named [`KeyvAdapter`](https://github.com/apollographql/apollo-utils/tree/main/packages/keyvAdapter#keyvadapter-class) in the `@apollo/utils.keyvadapter` package.

Reviewer 2: "provided" rather than "wrapped"?

Unfortunately, the default configuration of Apollo Server exposes you to denial of service attacks.
This is because APQs are enabled by default and the default cache in Apollo Server 3 is unbounded, meaning an attacker can exhaust your memory and crash your server. The default cache in Apollo Server 4 will be bounded; **we recommend opting in to that behavior by providing the `cache: "bounded"` option to your Apollo Server constructor or configuring the `cache` yourself.**

## Configuring in-memory caching

> **Apollo Server's default caching features use an unbounded cache, which is not safe for production use.** If you'd like to configure the in-memory cache, Apollo provides the [`InMemoryLRUCache`](https://github.com/apollographql/apollo-utils/tree/main/packages/keyValueCache#inmemorylrucache) class from the `@apollo/utils.keyvaluecache` package.

Reviewer 3: ```suggestion
**Apollo Server's default caching features use an unbounded cache, which is not safe for production use.** If you want to configure the in-memory cache, Apollo provides the [`InMemoryLRUCache`](https://github.com/apollographql/apollo-utils/tree/main/packages/keyValueCache#inmemorylrucache) class from the `@apollo/utils.keyvaluecache` package.
```

The `InMemoryLRUCache` class is a wrapper around the `lru-cache` package and has a default maximum of approximately 30MiB of memory. You can configure an instance of `InMemoryLRUCache` with the same options as found in the `lru-cache` package, see the [`lru-cache` documentation](https://www.npmjs.com/package/lru-cache) for more details. You can use Apollo's `InMemoryLRUCache` by passing it to the `cache` option of the `ApolloServer` constructor like so:

```ts
import { InMemoryLRUCache } from '@apollo/utils.keyvaluecache';

const server = new ApolloServer({
  // Note: this is equivalent to providing `cache: "bounded"`
  cache: new InMemoryLRUCache(),
});
```

> 

Reviewer 2: Maybe add a paragraph saying something like

As a shorthand, you can write `cache: "bounded"`; this is equivalent to `cache: new InMemoryLRUCache()` and doesn't require you to `npm install @apollo/utils.keyvaluecache` yourself.

In this example, we've increased the default size and provided a default TTL. For more information on these configuration options, see the [`lru-cache` documentation`](https://www.npmjs.com/package/lru-cache).
```ts
import { InMemoryLRUCache } from '@apollo/utils.keyvaluecache';

const server = new ApolloServer({
  // ...
  cache: new InMemoryLRUCache({
    // ~100MiB
    maxSize: Math.pow(2, 20) * 100,
    // 5 minutes (in milliseconds)
    ttl: 300_000,
  }),
});
```

## Configuring external caching

Apollo no longer maintains any caching backends directly. Instead, we recommend using the [`keyv`](https://www.npmjs.com/package/keyv) package along with the [`KeyvAdapter`](https://github.com/apollographql/apollo-utils/tree/main/packages/keyvAdapter#keyvadapter-class) class provided by the `@apollo/utils.keyvadapter` package. `KeyvAdapter` simply wraps a `Keyv` instance and implements the `KeyValueCache` interface which is required by Apollo Server. You can use the `KeyvAdapter` class to wrap a `Keyv` instance and provide it to the `cache` option of the `ApolloServer` constructor like so:

> ```ts

Reviewer 2: I do think we should provide complete examples for Redis and Memcached (replacing the old ones we have) which should include `npm install` commands: not obvious that you need to `npm install @keyv/redis`, eg.

Author: Let me know what you think: [909a913](https://github.com/apollographql/apollo-server/pull/6547/commits/909a913281eb58f7751178c14d51f72dff3d8ed0)

import Keyv from 'keyv';
import { KeyvAdapter } from '@apollo/utils.keyvadapter';

const server = new ApolloServer({
  // ...,
  cache: new KeyvAdapter(new Keyv('redis://localhost:6379')),
});
```

## Implementing your own cache backend

If your requirements are specialized or you'd prefer to implement your own cache backend, you can implement the `KeyValueCache` interface and pass it to the `ApolloServer` constructor directly.

The `KeyValueCache` interface is shown below:
```ts
interface KeyValueCache<V> {
  get(key: string): Promise<V | undefined>;
  // ttl is specified in seconds
  set(key: string, value: V, options?: { ttl?: number | null }): Promise<void>;
  delete(key: string): Promise<boolean | void>;
}
```

## Legacy caching implementation

> Before Apollo Server 3.9+, we used the `apollo-server-caching` package to implement caching. The `apollo-server-caching` package is no longer maintained, and we do not recommend using it. The `KeyValueCache` interface has been moved and is now in the `@apollo/utils.keyvaluecache` package.

Reviewer 3: ```suggestion
Versions of Apollo Server prior to 3.9 use the `apollo-server-caching` package to implement caching. The `apollo-server-caching` package is no longer maintained, and we do not recommend using it. The `KeyValueCache` interface has been moved and is now in the `@apollo/utils.keyvaluecache` package.
```

The `InMemoryLRUCache` class has also moved to the `@apollo/utils.keyvaluecache` package. The `InMemoryLRUCache` class now uses version 7 of `lru-cache`, accepting different configuration options and no longer allowing a cache to be unbounded.

> The `apollo-server-cache-redis` and `apollo-server-cache-memcached` packages are no longer published; we recommend using `keyv` instead.

Reviewer 2: `, as shown above`.

Author: [d8cbfe5](https://github.com/apollographql/apollo-server/pull/6547/commits/d8cbfe5e64d31f966e389439f6a795dc91745162)

> The `apollo-server-cache-redis` and `apollo-server-cache-memcached` packages are no longer published; we recommend using `keyv` instead.

Reviewer 2: maybe replace `are no longer published` with `are no longer receiving updates`, I dunno.

Author: [d8cbfe5](https://github.com/apollographql/apollo-server/pull/6547/commits/d8cbfe5e64d31f966e389439f6a795dc91745162)
